### PR TITLE
Visitor can register an account

### PR DESCRIPTION
**Pull request by Fraseer**
PT Link: https://www.pivotaltracker.com/story/show/179887318

## User Story
As a visitor
In order to read full articles
I would like to be able to register an account

## Changes Proposed
- Create and format fixtures for stubbing
- Create spec file for registering an account
- Create a register form
- Add I18N translations for all new language

## Screenshots
image

Skärmavbild 2021-10-26 kl  14 51 04

image

image

Ycu5jPUa8F

### CODE REVIEW

File: cypress/integration/userCanRegisterForAnAccount.feature.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    });
+    it("is expected to display a success message", () => {
+      cy.get("[data-cy=registration-message]").should(
+        "contain",
+        "Registration successful"
+      );
+    });
+    it("is expected to clear the input fields after registration", () => {
+      cy.get("body").trigger("keydown", { keyCode: 27 });
+      cy.wait(500);
```
Comment (dorianbuck) [inline cypress/integration/userCanRegisterForAnAccount.feature.js]: Is this necessary?

File: src/modules/article.js [hunk 2]
```diff
@@ -29,19 +30,19 @@ const Article = {
...
+//       payload:
+//         "We are sorry! Your request can not be processed at this time. Try again later",
+//     });
+//   } else {
+//     store.dispatch({
+//       type: "ERROR_MESSAGE",
+//       payload: error.message,
+//     });
+//   }
+// };
```
Comment (dorianbuck) [inline src/modules/article.js]: Is this comments?

File: src/state/store/initialState.js [hunk 3]
```diff
@@ -3,6 +3,8 @@ const initialState = {
   article: null,
   categories: [],
   authenticated: false,
+  error: false,
+  message: "",
```
Comment (dorianbuck) [inline src/state/store/initialState.js:7]: 👍🏽

File: src/components/Header.jsx [hunk 4]
```diff
@@ -49,16 +49,25 @@ const Header = () => {
...
+          <>
+            <Menu.Item position="right">
+              <Select
+                data-cy="language-selector"
+                placeholder={t("chooseLanguage")}
+                options={languageOptions}
+                onChange={(event, data) => {
+                  i18n.changeLanguage(data.value);
+                }}
+              ></Select>
```
Comment (1CIM) [inline src/components/Header.jsx]: Can this not be self-closing?

File: src/components/Header.jsx [hunk 5]
```diff
@@ -49,16 +49,25 @@ const Header = () => {
...
+            </Menu.Item>
+            <Menu.Item
+              id="sign-up"
+              name={t("signUp")}
+              as={Link}
+              to={{ pathname: "/register" }}
+              data-cy="sign-up-button"
+            ></Menu.Item>
```
Comment (1CIM) [inline src/components/Header.jsx]: Can this not be self-closing?

File: src/modules/article.js (hunk 2 above)
Comment (1CIM) [inline src/modules/article.js]: Why is this commented out code here?

**Review by 1CIM — CHANGES_REQUESTED** [review]
Left some comments

File: cypress/integration/userCanRegisterForAnAccount.feature.js (hunk 1 above)
Comment (maxarvid) [reply cypress/integration/userCanRegisterForAnAccount.feature.js]: It is not! Removed.

File: src/modules/article.js (hunk 2 above)
Comment (maxarvid) [reply src/modules/article.js]: Good catch, removed.

File: src/components/Header.jsx (hunk 4 above)
Comment (maxarvid) [reply src/components/Header.jsx]: It can and does look better. Fixed.

File: src/components/Header.jsx (hunk 5 above)
Comment (maxarvid) [reply src/components/Header.jsx]: Same as above, fixed.

File: src/modules/article.js (hunk 2 above)
Comment (maxarvid) [reply src/modules/article.js]: Erroneously left in, removed now.

**Review by 1CIM — APPROVED** [review]
👍

File: src/components/Register.jsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  const { error, authenticated, message } = useSelector((store) => store);
+  const [open, setOpen] = useState(false);
+  const { t } = useTranslation();
+
+  const handleSubmit = (event) => {
+    Authentication.register(event).then(setOpen(true));
+    document.getElementById("form-input-control-error-email").value = "";
+    document.getElementById("form-input-password").value = "";
+    document.getElementById("form-input-confirm-password").value = "";
+  };
```
Comment (tochman) [inline src/components/Register.jsx:24]: We can get the values of the child elements in a different way. I can pull down this code and show you if you want.

File: src/modules/authentication.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  async register(event) {
+    try {
+      await axios({
+        method: "post",
+        url: "/api/auth",
+        params: {
+          email: event.target.email.value,
+          password: event.target.password.value,
+          password_confirmation: event.target.confirmPassword.value,
+        },
```
Comment (tochman) [inline src/modules/authentication.js:15]: We need to use the authentication module in order to get the authentication headers

**Pull request merged**